diff -up waflib/Build.py.orig pycairo-1.10.0/.waf3-1.6.4-e3c1e08604b18a10567cfcd2d02eb6e6/waflib/Build.py
--- .waf3-1.6.4-e3c1e08604b18a10567cfcd2d02eb6e6/waflib/Build.py.orig	2015-11-03 19:40:47.691969757 +0100
+++ .waf3-1.6.4-e3c1e08604b18a10567cfcd2d02eb6e6/waflib/Build.py	2015-11-03 19:41:07.888153441 +0100
@@ -151,6 +151,7 @@ class BuildContext(Context.Context):
 				f.close()
 		self.init_dirs()
 	def store(self):
+		return
 		data={}
 		for x in SAVED_ATTRS:
 			data[x]=getattr(self,x)
