diff --git a/monitor/proxy/gproxyvolumemonitor.c b/monitor/proxy/gproxyvolumemonitor.c
index 7f05b39..c520cf9 100644
--- a/monitor/proxy/gproxyvolumemonitor.c
+++ b/monitor/proxy/gproxyvolumemonitor.c
@@ -1164,8 +1164,6 @@ is_remote_monitor_supported (const char *dbus_name)
                                                           NULL,
                                                           &error))
     {
-      g_printerr ("invoking IsSupported() failed for remote volume monitor with dbus name %s:: %s (%s, %d)\n",
-                  dbus_name, error->message, g_quark_to_string (error->domain), error->code);
       g_error_free (error);
       goto out;
     }
